---
output:
  html_document: default
  pdf_document: 
    latex_engine: xelatex
---
########################################################################################################
### PACKAGE INSTALLATION ###

If you are new to R, you may need to install some of the dependencies:
```{r eval=FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
install.packages("BiocManager")
BiocManager::install()

library(BiocManager)
BiocManager::install(c("caret", "devtools", "data.table", "dplyr", "furrr", "future", "future.apply",
  "ggplot2", "grDevices", "gridExtra", "igraph", "irlba", "janitor", "Matrix", "methods", 
  "plotly", "plyr", "purrr", "randomForest", "rhdf5", "rtracklayer", "scales", "stats", "stringr", 
  "tibble", "tidyr", "umap", "utils"))
```

Next, install a few additional dependencies found on Github, including amethyst itself.

```{r eval=FALSE}
devtools::install_github("JinmiaoChenLab/Rphenograph")
devtools::install_github("KrishnaswamyLab/MAGIC/Rmagic")
devtools::install_github("lrylaarsdam/amethyst")
```

Now load libraries into R:

```{r message=FALSE, warning=FALSE}
library(amethyst)
library(data.table)
library(ggplot2)
library(dplyr)
library(tibble)
library(tidyr)
library(plyr)
library(future)
library(furrr)
library(purrr)
library(cowplot)
library(pheatmap)
```

########################################################################################################
### LOADING PRACTICE DATA ###

First, download the practice data. This vignette comes with site-level CpG methylation information from 50 human banked PBMCs. Download the h5 file and associated metadata with the following commands.

* Note: By default, data will download to the ~/Downloads folder. Change if a different directory is desired.

```{r, eval=FALSE}
download.file("https://adeylabopen.s3.us-west-2.amazonaws.com/amethyst/pbmc_vignette.h5", "~/Downloads/pbmc_vignette.h5", method = "curl") # Contains site-level methylation information for each cell
download.file("https://raw.githubusercontent.com/lrylaarsdam/amethyst/main/pbmc_vignette/pbmc_vignette_cellInfo.txt", "~/Downloads/pbmc_vignette_cellInfo.txt") # Summary QC statistics for each cell
download.file("https://raw.githubusercontent.com/lrylaarsdam/amethyst/main/pbmc_vignette/pbmc_vignette.annot", "~/Downloads/pbmc_vignette.annot") # Simulated batch metadata
```

```{r, include=FALSE}
# Loading local workspace to speed up each knitr iteration
load("~/Library/CloudStorage/OneDrive-OregonHealth&ScienceUniversity/amethyst/github/pbmc_vignette/pbmc_vignette_workspace.RData")
```

########################################################################################################
### ASSEMBLING THE AMETHYST OBJECT ###

Now construct an amethyst object, which stores the path to the h5 file as well as a lot of other information that will be calculated downstream.

```{r, eval = FALSE}
obj <- createObject()
```

Next, we need to add metadata about each cell. Useful metadata includes quality control metrics contained in the .cellInfo.txt intermediate output or .annot files, if using the Adey lab Premethyst workflow.

```{r, eval = FALSE}
obj <- addCellInfo(obj, file = "~/Downloads/pbmc_vignette_cellInfo.txt")
obj <- addAnnot(obj, file = "~/Downloads/pbmc_vignette.annot", name = "batch") 
head(obj@metadata)
```

While not essential, it can be helpful to filter cells with outlying coverage values right away so downstream functions don't perform calculations for cells that will not be used. This can easily be done with dplyr logic. First, view the coverage distribution, then filter as necessary.

* Note: vignette data has been pre-filtered. We recommend cells have a minimum of 1M cytosines covered.

```{r, fig.width=5, fig.height=4}
ggplot(obj@metadata, aes(x = cov)) + geom_histogram(bins = 10) 
obj@metadata <- obj@metadata |> dplyr::filter(cov > 100000 & cov < 40000000)
```

Next, we need to specify the location of the h5 file containing site-level methylation data for each barcode. In this case, every barcode belongs to the same h5 file, but an unlimited number of h5 files can be used in the same object. Row names are barcodes and the "paths" column specifies each path.

```{r}
obj@h5paths <- data.frame(row.names = rownames(obj@metadata), paths = rep("~/Downloads/pbmc_vignette.h5", length(rownames(obj@metadata))))
head(obj@h5paths)
```

########################################################################################################
### CLUSTERING ###

The next step is to cluster cells, which we typically do based on methylation values over fixed genomic windows. 
An initial indexing helps reduce the computational load by determining the locations corresponding to each chromosome in every h5 file and only calculating across one at a time.

* Note: You may have to copy/paste any code reading the h5 file (e.g., makeWindows) directly into the console instead of running the chunk, if using the .Rmd template

```{r message=FALSE, warning=FALSE, eval = FALSE}
obj@index[["chr_cg"]] <- indexChr(obj, type = "CG", threads = 1) 
obj@genomeMatrices[["cg_100k_score"]] <- makeWindows(obj,
                                                     stepsize = 100000, 
                                                     type = "CG", 
                                                     metric = "score", 
                                                     threads = 1, 
                                                     index = "chr_cg", 
                                                     nmin = 2) 
```

You may want to remove windows where many values are NA. The appropriate threshold will highly depend on how big the windows are and how many cells you have.
In this case, since the vignette data is high coverage and the genomic windows are large, I am going to filter for at least 90% of the cells have values in that window.

```{r, eval = FALSE}
obj@genomeMatrices[["cg_100k_score"]] <- obj@genomeMatrices[["cg_100k_score"]][rowSums(!is.na(obj@genomeMatrices[["cg_100k_score"]])) >= nrow(obj@metadata)*.9, ]
```

Next, perform dimensionality reduction with irlba. If you are unsure how many dimensions to use, the dimEstimate function can estimate the number needed to explain the desired variance threshold.
*Note: In this example, the number of requested output dimensions is low because pbmc_vignette.h5 has 50 cells. 

```{r}
dimEstimate(obj, genomeMatrices = c("cg_100k_score"), dims = c(10), threshold = 0.95)
```

As suggested, we will reduce the data from obj@genomeMatrices[["cg_100k_score"]] into seven dimensions using the irlba package, which performs fast truncated singular value decomposition.

```{r}
set.seed(111)
obj@reductions[["irlba"]] <- runIrlba(obj, genomeMatrices = c("cg_100k_score"), dims = c(7), replaceNA = c(0))
obj@reductions[["irlba_regressed"]] <- regressCovBias(obj, reduction = "irlba") # Optional; helps reduce coverage bias in clustering
```

Now determine cluster membership using the Rphenograph package.
*Note: In this example, k_phenograph and neighbors parameters are low because pbmc_vignette.h5 has 50 cells.

```{r}
set.seed(111)
obj <- runCluster(obj, k_phenograph = 10, reduction = "irlba_regressed") # consider increasing k_phenograph to 50 for larger datasets
```

Umap and tsne are dimensionality reduction techniques frequently used to collapse higher-order information into two or three
coordinates for visualization of single-cell data. Either or both methods can be used as input for dimensionality reduction-
based visualization functions like dimFeature. 

```{r}
set.seed(111)
obj <- runUmap(obj, neighbors = 5, dist = 0.05, method = "euclidean", reduction = "irlba_regressed") 
obj <- runTsne(obj, perplexity = 10, method = "euclidean", theta = 0.2, reduction = "irlba_regressed") 
```

### Visualizing the results ###
First, plot the UMAP or TSNE coordinates of the cells with the color corresponding to cluster membership.

```{r, fig.width=10, fig.height=4}
p1 <- dimFeature(obj, colorBy = cluster_id, reduction = "umap") + ggtitle("UMAP")
p2 <- dimFeature(obj, colorBy = cluster_id, reduction = "tsne") + ggtitle("TSNE")
plot_grid(p1, p2)
```

You might find that fixed genomic windows don't give you good resolution of groups. Any feature set can be used for dimensionality reduction input.
The makeWindows function can also calculate methylation levels over a bed file or genes (not recommended unless you are calculating %mCH.)
Here is another clustering example using a set of pre-identified PBMC differentially methylated regions (DMRs).

```{r, message=FALSE, eval = FALSE}
download.file("https://raw.githubusercontent.com/lrylaarsdam/amethyst/main/pbmc_vignette/pbmc_highconfidence_dmrs.bed", "~/Downloads/pbmc_dmr.bed") 
obj@genomeMatrices[["pbmc_dmrs"]] <- makeWindows(obj, bed = "~/Downloads/pbmc_dmr.bed", type = "CG", metric = "percent", threads = 1, index = "chr_cg", nmin = 2)
obj@genomeMatrices[["pbmc_dmrs"]] <- obj@genomeMatrices[["pbmc_dmrs"]][rowSums(!is.na(obj@genomeMatrices[["pbmc_dmrs"]])) >= nrow(obj@metadata)*.2, ] 
```

Now re-run irlba with the DMR-based windows:

```{r}
dimEstimate(obj, genomeMatrices = c("pbmc_dmrs"), dims = c(10), threshold = 0.90)
set.seed(111)
obj@reductions[["irlba"]] <- runIrlba(obj, genomeMatrices = c("pbmc_dmrs"), dims = c(8), replaceNA = c(0))
obj@reductions[["irlba_regressed"]] <- regressCovBias(obj, reduction = "irlba")
```

Now re-run clustering and UMAP:

```{r, fig.width=5, fig.height=4}
set.seed(111)
obj <- runCluster(obj, k_phenograph = 10, reduction = "irlba_regressed") # consider increasing k_phenograph to 50 for larger datasets
set.seed(111)
obj <- runUmap(obj, neighbors = 5, dist = 0.05, method = "euclidean", reduction = "irlba_regressed") 
dimFeature(obj, colorBy = cluster_id, reduction = "umap")
```

In this context, DMRs worked much better than genomic windows for clustering.
dimFeature uses ggplot logic, so you can easily modify plots as needed. For example:

```{r, fig.width=10, fig.height=4}
dimFeature(obj, colorBy = cluster_id) + facet_wrap(vars(batch)) # Batch is simulated to illustrate function utility. Any column in the metadata will work.
```

(Optional) If you want to make the umap/tsne plots look nicer, amethyst provides many built-in color palettes:

```{r, fig.width=3, fig.height=5}
testPalette(output = "swatch", n = length(unique(obj@metadata$cluster_id)))
```

```{r, fig.width=5, fig.height=4}
pal <- c("#F9AB60", "#E7576E", "#630661", "#B5DCA5") # makePalette(option = 7, n = 4) 
dimFeature(obj, colorBy = cluster_id, colors = pal, pointSize = 1)
```

Show the distribution of cluster membership between samples with sampleComp. Plots can be easily modified with ggplot command logic.

```{r, fig.width=5, fig.height=5}
sampleComp(obj, groupBy = "batch", colorBy = "cluster_id", colors = pal) 
```

dimFeature is useful for visualizing how different parameters in the cellInfo file are distributed throughout the UMAP:

```{r, fig.width=8, fig.height=4}
p1 <- dimFeature(obj, colorBy = log(cov), pointSize = 1) + scale_color_gradientn(colors = c("black", "turquoise", "gold", "red")) + ggtitle("Coverage distribution")
p2 <- dimFeature(obj, colorBy = mcg_pct, pointSize = 1) + scale_color_gradientn(colors = c("black", "turquoise", "gold", "red")) + ggtitle("Global %mCG distribution")
plot_grid(p1, p2)
```

########################################################################################################
### ANNOTATION ###

Now that we have clusters, the next step is annotation. There are a couple ways to do this:

1) One useful method is to look at mCG hypomethylation over canonical marker genes. We recommend visualizing mCG patterns over the entire gene body. 
The first step is to load an annotation file for the reference genome so amethyst knows the coordinates for each gene.

```{r, eval = FALSE}
obj@ref <- makeRef("hg38")
```

Next, calculate methylation levels in short genomic windows for each cluster. We recommend 500bp windows, but 1kb are used here since the dataset is smaller.

```{r, message=FALSE, warning=FALSE, eval = FALSE}
cluster1kbwindows <- calcSmoothedWindows(obj, 
                                         type = "CG", 
                                         threads = 1,
                                         step = 1000,
                                         smooth = 3,
                                         genome = "hg38",
                                         index = "chr_cg",
                                         groupBy = "cluster_id",
                                         returnSumMatrix = TRUE, # save sum matrix for DMR analysis
                                         returnPctMatrix = TRUE)
obj@genomeMatrices[["cg_cluster_tracks"]] <- cluster1kbwindows[["pct_matrix"]]
```

Now you can view methylation patterns over key marker genes:

```{r, fig.width=15, fig.height=10}
heatMap(obj, 
        genes = c("SPI1", "CD2", "S100A8",  "CD79A", "CD3G", "ELANE", "MPO", 
           "S100A8", "MPEG1", "IRF8", "CD74", "GZMK", "CD3E", "CD3D", "KIR2DL4", "KLRB1"), 
        matrix = "cg_cluster_tracks", 
        nrow = 4,
        legend = FALSE,
        width = 1000)
```

As you can see from the heatMaps, promoters are often universally hypomethylated or not at the predicted site. 
Because of this it is often more informative to look at mCG over the entire gene body.
In addition to heatMap, this can also be done with the histograM function:

```{r, fig.width=15, fig.height=7}
histograM(obj, genes = c("ELANE", "MPEG1", "SPI1", "CD2", "CD3D"), matrix = "cg_cluster_tracks", legend = F, width = 1000)
```

However, it can still be useful to look at aggregated promoter metrics. 
To calculate %mCG for promoter regions (in this example all protein coding genes, but a subset can easily be used):

```{r, message=FALSE, eval = FALSE}
protein_coding <- unique(obj@ref |> dplyr::filter(type == "gene" & gene_type == "protein_coding" & seqid != "chrM") |> dplyr::pull(gene_name))
obj@genomeMatrices[["cg_promoter"]] <- makeWindows(obj, 
                                                     genes = protein_coding,
                                                     promoter = TRUE, 
                                                     type = "CG", 
                                                     metric = "percent", 
                                                     threads = 1, 
                                                     index = "chr_cg", 
                                                     nmin = 2) 
# subsetting to genes with values in at least 5 cells
obj@genomeMatrices[["cg_promoter"]] <- obj@genomeMatrices[["cg_promoter"]][rowSums(!is.na(obj@genomeMatrices[["cg_promoter"]])) >= 10, ]
```

Now you can view average %mCG of marker gene promoters by cluster:

```{r, fig.height=3, fig.width=9, message=FALSE}
genes <- c("SPI1", "CD19", "CD2", "CD6", "CD8A", "CD4", "CSF1R", "GATA1", "CD79A", "CD3G", "ELANE", "MPO", "ITGAM",
           "S100A8", "MPEG1", "FN1", "IRF8", "CD74", "RORA", "GZMK", "CD3E", "CD3D", "MEIS1", "KIR2DL4")
dotM(obj, genes = genes, groupBy = "cluster_id", matrix = "cg_promoter")
```

Again, easily modify with ggplot logic as desired: 

```{r, fig.height=3, fig.width=9, message=FALSE}
dotM(obj, genes = genes, groupBy = "cluster_id", matrix = "cg_promoter") + 
  scale_color_gradientn(colors =  c("#FF0082", "#dbdbdb", "#cccccc", "#999999")) + scale_size(range = c(1, 8))
```

It can also be helpful to use a less directed approach when determining differences between groups. 
For this vignette, we are just testing a subset of known marker genes, but for thorough data analysis it would better to test all protein coding genes.

```{r fig.height=4, fig.width=4, message=FALSE, warning=FALSE}
cluster_promoter_markers <- findClusterMarkers(obj, 
                                               matrix = "cg_promoter", 
                                               genes = genes, 
                                               threads = 1)
cluster_promoter_markers <- cluster_promoter_markers |> dplyr::filter(p.adj < 0.05) # Not many results because it's a small dataset
head(cluster_promoter_markers)
```

```{r fig.height=4, fig.width=4, message=FALSE, warning=FALSE}
dotM(obj, genes = cluster_promoter_markers$gene, groupBy = "cluster_id", matrix = "cg_promoter") + 
  scale_color_gradientn(colors =  c("#FF0082", "#dbdbdb", "#cccccc", "#999999")) + scale_size(range = c(1, 12))
```

Another method one could use is by comparison to an annotated reference. 
While few exist, we have put aggregated methylation levels per group over high-confidence PBMC DMRs on Github. 
First download this data and calculate average methylation levels per cluster for each DMR (windows or any feature can also work)

```{r}
download.file("https://raw.githubusercontent.com/lrylaarsdam/amethyst/main/pbmc_vignette/pbmc_ref.RData", "~/Downloads/pbmc_ref.RData") 
ref <- readRDS("~/Downloads/pbmc_ref.RData")

obj@genomeMatrices[["pbmc_dmrs_aggregated"]] <- aggregateMatrix(obj, matrix = "pbmc_dmrs", groupBy = "cluster_id")
```

Now view correlation profiles

```{r fig.height=4, fig.width=4}
cor <- cor(merge(ref, 
                 obj@genomeMatrices[["pbmc_dmrs_aggregated"]], 
                 by = 0) |> tibble::column_to_rownames(var = "Row.names"), use = "pairwise.complete.obs")
cor <- cor[c(1:ncol(ref)), c((ncol(ref) + 1)):ncol(cor)]
pheatmap(cor)
```

Based on all these annotation tools, we can rename our clusters according to broad class using dplyr logic:

```{r, fig.height=4, fig.width=5}
obj@metadata[["type"]] <- dplyr::recode(obj@metadata[["cluster_id"]],
                                             "1" = "T", 
                                             "2" = "NK", 
                                             "3" = "B",
                                             "4" = "Mono")
library(ggrepel)
dimFeature(obj, colorBy = type, colors = pal, pointSize = 1)  +
  geom_text_repel(aes(label = type), color = "black", data = obj@metadata |> dplyr::group_by(type) |> dplyr::summarise(umap_x = median(umap_x), umap_y = median(umap_y)))
```

You might also want cluster tracks with the group name:

```{r, fig.height=10, fig.width=15}
obj@genomeMatrices[["cg_type_tracks"]] <- copy(obj@genomeMatrices[["cg_cluster_tracks"]])
setnames(obj@genomeMatrices[["cg_type_tracks"]], c("chr", "start", "end", "T", "NK", "B", "Mono"))
heatMap(obj, 
        genes = c("SPI1", "CD2", "S100A8",  "CD79A", "CD3G", "ELANE", "MPO", 
           "S100A8", "MPEG1", "IRF8", "CD74", "GZMK", "CD3E", "CD3D", "KIR2DL4", "KLRB1"), 
        matrix = "cg_type_tracks", 
        nrow = 4,
        legend = FALSE,
        width = 1000)
```

########################################################################################################
### DIFFERENTIALLY METHYLATED REGION ANALYSIS ###

There are two main formats to set up DMR analysis. The first is to test DMRs for each cluster against all others.
Only the sum matrix (which we saved at the calcSmoothedWindows step) is needed, or regenerate with your annotated cell types:

```{r}
dmrs <- testDMR(cluster1kbwindows[["sum_matrix"]], eachVsAll = TRUE, nminTotal = 5, nminGroup = 5) # or use cluster1kbwindows[["sum_matrix"]] and rename
```

Then expand and filter the resulting list according to the desired stringency.

```{r}
dmrs <- filterDMR(dmrs, method = "bonferroni", filter = TRUE, pThreshold = 0.01, logThreshold = 2)
head(dmrs)
```

Especially since the matrix is smoothed, adjacent genomic windows may be significant. You can collapse them with the following function.
If annotation = T, any overlapping genes will be noted in the results table.

```{r}
collapsed_dmrs <- collapseDMR(obj, dmrs, maxDist = 4000, minLength = 2000, reduce = T, annotate = T) 
head(collapsed_dmrs)
```

Note: The "test" column indicates which cluster is considered the member group. If you are testing a renamed matrix, you might want to add those names
To your results instead of having the numerical order in which they were tested (which works ok for cluster DMRs). Change sumMatrix if necessary.

```{r}
key <- data.frame(test = as.factor(1:4), 
                type = c("T", "NK", "B", "Mono"))
collapsed_dmrs <- left_join(collapsed_dmrs, key, by = "test")
head(collapsed_dmrs)
```

If specific comparisons are desired, a data frame can be provided describing the tests. Three columns should be included: One listing members of group A, one listing members of group B, and one with the name of the test.

```{r, eval=FALSE}
comparisons <- data.frame(
  stringsAsFactors = FALSE,
              name = c("test1", "test2", "test3"),
                 A = c("1,2,3", "1", "2,3"),
                 B = c("1,4", "2", "1")
)
dmrs <- testDMR(sumMatrix = cluster1kbwindows[["sum_matrix"]], comparisons = comparisons, nminTotal = 5, nminGroup = 5)
```

########################################################################################################
### INVESTIGATING DMR RESULTS ### 

First, let's look at how many DMRs were identified in each group:

```{r, message=FALSE, fig.height=5, fig.width=5}
ggplot(collapsed_dmrs |> dplyr::group_by(type, direction) |> dplyr::summarise(n = n()), 
       aes(y = type, x = n, fill = type)) + geom_col() + 
  facet_grid(vars(direction), scales = "free_y") + scale_fill_manual(values = pal) + theme_classic()
```

Isolate top results per group 
We find it helpful to select by a combined metric of logFC and padj, but you can modify as necessary:

```{r}
top_dmrs <- collapsed_dmrs |> 
  dplyr::group_by(type, direction) |> 
  dplyr::arrange(dmr_padj, .by_group = TRUE) |> dplyr::mutate(rank_padj = 1:n()) |>
  dplyr::arrange(desc(abs(dmr_logFC)), .by_group = TRUE) |> dplyr::mutate(rank_logFC = 1:n()) |>
  rowwise() |> dplyr::mutate(total_rank = sum(rank_padj, rank_logFC)) |> 
  group_by(test, direction) |> slice_min(n = 1, order_by = total_rank) |>
  dplyr::mutate(location = paste0(chr, "_", (dmr_start - 2000), "_", (dmr_end + 2000))) |> dplyr::arrange(direction)
```

Plotting top hypomethylated regions for each group shows expected patterns based on known marker genes:

```{r, fig.height=8, fig.width=15}
heatMap(obj, matrix = "cg_type_tracks", regions = top_dmrs$location[top_dmrs$direction == "hypo"], nrow = 2, legend = FALSE, width = 1000, arrowOverhang = 0)
```

########################################################################################################
### GO ANALYSIS ###

Further interpretation of the results can be explored using a wide variety of packages available on R.
In this example, we will use the topGO package to test for Gene Ontology (GO) term enrichments for genes with hypomethylated regions in the T cell group.

```{r, message=FALSE}
library(topGO)
background <- rownames(obj@genomeMatrices[["cg_promoter"]]) # all genes tested 
query <- unlist(strsplit(collapsed_dmrs$gene_names[collapsed_dmrs$type == "T" & collapsed_dmrs$direction == "hypo"], ", "))

GOdata <- new("topGOdata", 
              description = "GO Enrichment Analysis", 
              ontology = "BP", 
              allGenes = setNames(factor(as.integer(background %in% query), levels = c(0, 1)), background),
              geneSel = function(x) x == 1, 
              nodeSize = 10, 
              annot = annFUN.org, 
              mapping = "org.Hs.eg.db", 
              ID = "symbol")
resultElim <- runTest(GOdata, algorithm = "elim", statistic = "fisher")
resultElim <- GenTable(GOdata, Fisher = resultElim, topNodes = 500, numChar = 60)
resultElim <- resultElim |> dplyr::filter(Fisher < 0.01 & Significant > 5) |> dplyr::mutate(fold_change = Significant/Expected, Fisher = as.numeric(Fisher))
resultElim <- janitor::clean_names(resultElim)
```

As expected, top results are strongly related to T cell processes:

```{r}
ggplot(resultElim, aes(x = fold_change, y = reorder(term, fold_change), fill = fisher)) + geom_col() + theme_classic() + scale_fill_viridis_c(direction = -1)
```

########################################################################################################
### CONCLUSION ### 

Thanks for trying out Amethyst!
Additional utilities are still to come. We are also open to suggestions.
Good luck in your analysis!



